src/errors.rs

```diff
@@ -1,79 +1,152 @@
+    #[error(transparent)]
+    TomlEditParse(#[from] ::toml_edit::TomlError),
+    /// A ReqParseError from the semver crate
+    #[error(transparent)]
+    SemVerParse(#[from] semver::ReqParseError),
+    /// A SemVerError from the semver crate
+    #[error(transparent)]
+    SemVer(#[from] semver::SemVerError),
+    /// An error from the reqwest crate
+    #[error(transparent)]
+    Reqwest(#[from] ::reqwest::Error),
+
+    // links
+    /// Failed to read home directory
+    #[error("Failed to read home directory")]
+    ReadHomeDirFailure,
+
+    /// Invalid JSON in registry index
+    #[error("Invalid JSON in registry index")]
+    InvalidSummaryJson,
+
+    /// Given crate name is empty
+    #[error("Found empty crate name")]
+    EmptyCrateName,
+
+    /// No crate by that name exists
+    #[error("The crate `{0}` could not be found in registry index.")]
+    NoCrate(String),
+
+    /// No versions available
+    #[error("No available versions exist. Either all were yanked or only prerelease versions exist. Trying with the --allow-prerelease flag might solve the issue.")]
+    NoVersionsAvailable,
+
+    /// Unable to parse external Cargo.toml
+    #[error("Unable to parse external Cargo.toml")]
+    ParseCargoToml,
+
+    /// Cargo.toml could not be found.
+    #[error("Unable to find Cargo.toml")]
+    MissingManifest,
+
+    /// Cargo.toml is valid toml, but doesn't contain the expected fields
+    #[error("Cargo.toml missing expected `package` or `project` fields")]
+    InvalidManifest,
+
+    /// Found a workspace manifest when expecting a normal manifest
+    #[error("Found virtual manifest, but this command requires running against an actual package in this workspace.")]
+    UnexpectedRootManifest,
+
+    /// The TOML table could not be found.
+    #[error("The table `{0}` could not be found.")]
+    NonExistentTable(String),
+
+    /// The dependency could not be found.
+    #[error("The dependency `{name}` could not be found in `{table}`.")]
+    NonExistentDependency {
+        /// Name of the non-existent dependency
+        name: String,
+        /// Table of dependencies
+        table: String,
+    },
+
+    /// Config of cargo is invalid
+    #[error("Invalid cargo config")]
+    InvalidCargoConfig,
+
+    /// Unable to find the source specified by 'replace-with'
+    #[error("The source '{0}' could not be found")]
+    NoSuchSourceFound(String),
+
+    /// Unable to find the specified registry
+    #[error("The registry '{0}' could not be found")]
+    NoSuchRegistryFound(String),
+
+    /// Failed to parse a version for a dependency
+    #[error("The version `{version}` for the dependency `{dep}` couldn't be parsed")]
+    ParseVersion {
+        /// The invalid version
+        version: String,
+        /// The dependency with an invalid version
+        dep: String,
+    },
+
+    /// A string error
+    #[error("{0}")]
+    Custom(String),
+
+    /// Wraps another error in order to provide more context
```

Comment:
hmm, is it a common way of doing this thing?
I see that https://github.com/dtolnay/anyhow for example has a `context` function for this. Maybe `thiserror` has something similar?

Reply:
Unless I'm missing something, `thiserror` allows you to add `source` for an error variant by one of three ways:

1. prepending the field with `#[from]`,
2. prepending the field with `#[source]`, or
3. naming the field `source`

In all cases, the source of the error variant is kept in the variant itself, and I can't find any way to add this source information alongside the `Error` enum in the same way that `error-chain` did with the `Error`/`ErrorKind` setup. I could rename the enum to be `ErrorKind` and have an `Error` struct that stores the `ErrorKind` and a source, but then we lose the benefits of using `thiserror` since we won't be able to take advantage of the code generation that it does to allow us to convert those source errors to our `Error` enum.

Reply:
It just feels like a something is not right here, either:
1. `thiserror` is not the right tool for the job, or
2. we should redesign the error types exposed in the library

and I'll need more time to think about this

Reply:
I've got an idea about this that I'll try out today

Reply:
Thinking a little more about this, I think using a combination of `thiserror` and `anyhow` might be the right way to go. `thiserror` is good for defining the error types for the library and deriving the `Error` trait on them, and `anyhow` seems perfect for working with them within the application itself. I'll give that a try and see how it turns out

Reply:
any updates on this? (also needs merging with master)

Reply:
If we need to add some context to each error, maybe the error type should be
```rust
struct Error {
    pub error: ErrorKind,
    context: &'static str,
}
``` 
or something along the lines, or we could add more variants of errors to the lib.

Reply:
yea that's what I'm in the process of moving to. sorry I haven't been more responsive, I'm still working on this and will hopefully have an update for you tomorrow

Reply:
no worries :) take your time and thank you for the PR!